data/resources.go

```diff
@@ -9,104 +9,243 @@ import (
-	return int64(60 * 12 * float64(p.PopulationProdLevel) * math.Pow(1.1, float64(p.PopulationProdLevel)) * productivity)
+func silicaHourlyProductionRate(level int) int64 {
+	return genericHourlyProductionRate(level, 160, 1.05)
 }
 
-// getNewStock calculates the current value in stock for a resource based on value and duration since last update.
-func getNewStock(val int64, last time.Time, rate int64, storageLevel int, now time.Time) int64 {
-	duration := int64(now.Sub(last) / time.Second)
-	maxStorage := getMaxStorage(storageLevel)
-	const secondsPerHour = 60 * 60
-	current := val + duration*rate/secondsPerHour
-	if current >= maxStorage {
-		return maxStorage
-	}
-	return current
-}
-
-// UpdatePlanetResources updates the current planet struct to get up-to-date state
-func UpdatePlanetResources(p *ent.Planet, now time.Time) {
-	energyProductivity := getProductivityBasedOnEnergyConsumption(p)
-
-	p.Metal = getNewStock(
-		p.Metal,
-		p.LastResourceUpdate,
-		getMetalRate(p, energyProductivity),
-		p.MetalStorageLevel,
-		now,
-	)
-	p.Hydrogen = getNewStock(
-		p.Hydrogen,
-		p.LastResourceUpdate,
-		getHydrogenRate(p, energyProductivity),
-		p.HydrogenStorageLevel,
-		now,
-	)
-	p.Silica = getNewStock(
-		p.Silica,
-		p.LastResourceUpdate,
-		getSilicaRate(p, energyProductivity),
-		p.SilicaStorageLevel,
-		now,
-	)
-	p.Population = getNewStock(
-		p.Population,
-		p.LastResourceUpdate,
-		getPopulationRate(p, energyProductivity),
-		p.PopulationStorageLevel,
-		now,
-	)
-	p.LastResourceUpdate = now
+func genericStorageCapacity(level int, initial, base float64) int64 {
+	const expBase = 1.5
+	l := float64(level)
+	return int64(initial + base*l*l*math.Pow(expBase, l))
+}
+
+func metalStorageCapacity(level int) int64 {
+	return genericStorageCapacity(level, 1000, 1000)
+}
+
+func hydrogenStorageCapacity(level int) int64 {
+	return genericStorageCapacity(level, 1000, 1200)
+}
+
+func silicaStorageCapacity(level int) int64 {
+	return genericStorageCapacity(level, 1000, 1100)
+}
+
+func genericConstructionCost(level int, initial, base float64) int64 {
+	l := float64(level)
+	prevL := float64(level - 1)
+	upgradeCost := prevL * initial * (math.Pow(base, l) - math.Pow(base, prevL))
+	newCost := initial * math.Pow(base, l)
+	return int64(upgradeCost + newCost)
+}
+
+func genericPopulationEmployment(level int, initial, base float64) int64 {
+	l := float64(level)
+	return int64(initial * l * math.Pow(base, l))
+}
+
+func genericEnergyUsage(level int, initial, base float64) int64 {
+	l := float64(level)
+	return int64(initial * l * math.Pow(base, l))
+}
+
+func genericStorageCost(level int, initial, base float64) int64 {
+	l := float64(level)
+	return int64(initial * l * l * math.Pow(base, l))
+}
+
+func genericUpgradeDuration(level int, initial time.Duration, base float64) time.Duration {
+	l := float64(level)
+	i := float64(initial)
+	return time.Duration(i * math.Pow(base, l))
+}
+
+func metalMineUpgradeDuration(level int) time.Duration {
```

Comment:
can we not instead of having a wraper for each work with named macros like in C thus just calling generic function with proper macros as args everytime? just asking, don't know if its better

Reply:
I don't think Go has named macros. As far as I know, macros are used for optimization. If you really want you can figure out how [function inlining](https://github.com/golang/go/wiki/CompilerOptimizations#function-inlining) works in the Go compiler which optimizes it in a similar way.

Reply:
I considered putting the contents of this function directly in the place where it is called (in buildings.go). But I chose not to do this because having similar functions together makes it a lot easier to tweak the values with respect to eachother.